File: src/main/java/com/github/simiacryptus/aicoder/util/LineComment.java
package com.github.simiacryptus.aicoder.util;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.stream.Collectors;

public class LineComment extends IndentedText {
    public static class Factory implements TextBlockFactory<LineComment> {
        public final String commentPrefix;

        public Factory(String commentPrefix) {
            this.commentPrefix = commentPrefix;
        }

        @Override
        public @NotNull LineComment fromString(String text) {
            text = text.replace("\t", TAB_REPLACEMENT);
            CharSequence indent = StringTools.getWhitespacePrefix(text.split(DELIMITER));
            return new LineComment(commentPrefix, indent, Arrays.stream(text.split(DELIMITER))
                    .map(s -> StringTools.stripPrefix(s, indent))
                    .map(StringTools::trimPrefix)
                    .map(s -> StringTools.stripPrefix(s, commentPrefix))
                    .toArray(CharSequence[]::new));
        }

        @Override
        public boolean looksLike(@NotNull String text) {
            return Arrays.stream(text.split(DELIMITER)).allMatch(x->x.trim().startsWith(commentPrefix));
        }
    }

    private final CharSequence commentPrefix;

    public LineComment(CharSequence commentPrefix, CharSequence indent, CharSequence... textBlock) {
        super(indent, textBlock);
        this.commentPrefix = commentPrefix;
    }

    @Override
    public @NotNull String toString() {
        return commentPrefix + " " + Arrays.stream(rawString()).collect(Collectors.joining(DELIMITER + getIndent() + commentPrefix + " "));
    }

    public @NotNull LineComment withIndent(CharSequence indent) {
        return new LineComment(commentPrefix, indent, textBlock);
    }
}

File: src/main/kotlin/com/github/simiacryptus/aicoder/util/LineComment.kt
package com.github.simiacryptus.aicoder.util

import com.github.simiacryptus.aicoder.util.StringTools.getWhitespacePrefix
import com.github.simiacryptus.aicoder.util.StringTools.stripPrefix
import com.github.simiacryptus.aicoder.util.StringTools.trimPrefix
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors

class LineComment(private val commentPrefix: CharSequence, indent: CharSequence?, vararg lines: CharSequence) :
    IndentedText(indent!!, *lines) {
    class Factory(val commentPrefix: String) : TextBlockFactory<LineComment?> {
        override fun fromString(text: String?): LineComment {
            var text = text
            text = text!!.replace(Regex("\t"), TextBlock.TAB_REPLACEMENT.toString())
            val indent = getWhitespacePrefix(*text.split(TextBlock.DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray())
            return LineComment(
                commentPrefix,
                indent,
                *Arrays.stream(text.split(TextBlock.DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }
                    .toTypedArray())
                    .map { s: String? ->
                        stripPrefix(
                            s!!, indent
                        )
                    }
                    .map(Function<CharSequence, CharSequence> { obj: CharSequence -> trimPrefix(obj) })
                    .map { s: CharSequence? ->
                        stripPrefix(
                            s!!,
                            commentPrefix
                        )
                    }
                    .collect(Collectors.toList()).toTypedArray())
        }

        override fun looksLike(text: String?): Boolean {
            return Arrays.stream(text!!.split(TextBlock.DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }
                .toTypedArray()).allMatch { x: String ->
                x.trim { it <= ' ' }.startsWith(
                    commentPrefix
                )
            }
        }
    }

    override fun toString(): String {
        return commentPrefix.toString() + " " + Arrays.stream(rawString())
            .collect(Collectors.joining(TextBlock.DELIMITER + indent + commentPrefix + " "))
    }

    override fun withIndent(indent: CharSequence?): LineComment {
        return LineComment(commentPrefix, indent, *lines)
    }
}